Added Old Vestel A/C support (56 Bits) with full functions.

EUA opened the pull request:
Complete Vestel A/C support (56 Bits)

EUA committed:
Added Old Vestel A/C support (56 Bits) with full functions.

EUA committed:
Added travis support.

EUA committed:
fixed keywords file.

crankyoldgit commented:
First off, wow. Thanks for tackling adding a full on send/decode/class for this new protocol.

I'll start a code review in a minute.
The Travis build has picked up a number of issues that we'll need to address before we can merge.
I'll try to limit the feedback to just getting this merge-worthy and do some additional cleanup (unit tests and integrate with the examples in a subsequent PR)

crankyoldgit commented on src/IRsend.h:
```diff
@@ -308,6 +308,9 @@ class IRsend {
                uint16_t repeat = kNoRepeat);
 #endif
+#if SEND_VESTEL_AC
+  void sendVestelAC(uint64_t data, uint16_t nbits = kMideaBits);
```
I think you meant `kVestelACBits`.

EUA replied:
Ops. Residue from template file.

EUA commented:
Fixing travis issues now...

EUA committed:
Formatted some code lines.
Removed unused variable.

crankyoldgit commented on src/ir_Vestel.cpp:
```diff
@@ -0,0 +1,335 @@
+void IRsend::sendVestelAC(uint64_t data, uint16_t nbits) {
+  if (nbits % 8 != 0) return;  // nbits is required to be a multiple of 8.
+
+  // Set IR carrier frequency
```
You should be able to send this all in a single [sendGeneric()](https://github.com/markszabo/IRremoteESP8266/blob/master/src/IRsend.cpp#L333) call.
e.g.
```
sendGeneric(
    kVestelACHdrMark, kVestelACHdrSpace,  // Header
    kVestelACBitMark, kVestelACOneSpace,  // Data
    kVestelACBitMark,  kVestelACZeroSpace,
    kVestelACBitMark, 100000,  // Footer + repeat gap
    data, nbits, 38, false, repeat, 50);
```

EUA replied:
Ok. This looks working too.

crankyoldgit commented on src/ir_Vestel.cpp:
```diff
@@ -0,0 +1,335 @@
+//
+// Status: STABLE / Working.
+//
+void IRsend::sendVestelAC(uint64_t data, uint16_t nbits) {
```
You're missing the `uint16_t repeat` argument in almost all other protocols.
e.g. https://github.com/markszabo/IRremoteESP8266/blob/master/src/ir_Midea.cpp#L55

EUA replied:
I thought **repeat** is some remotes mandatory spec.
Like if receiver expects same code twice to accept instead of CRC.

crankyoldgit replied:
Some protocols a `repeat` doesn't happen or makes little/no sense. That's probably the case here. However, it's part of the calling API is that all _real_ protocols support it.
e.g. Sony's protocol requires a min of 3 repeats. Sometimes it's used to simulate the button being held. Anyway, sendGeneric supports it, and you can use a default value in the header file so it can be invoked without it.

crankyoldgit commented on src/ir_Vestel.cpp:
```diff
@@ -0,0 +1,335 @@
+  //Just counts the set bits +1 on stream and take inverse after mask
+  uint8_t sum = 0;
+  uint64_t temp_state = state & kVestelACCRCMask;
+  for (uint8_t i = 0/*+(8+8+4)*/; i < 64; i++) {
```
The following should be a faster way to count the bits as it can shortcut when there are no more bits to count. i.e. It doesn't have to loop 64 times.
```
    for (; temp_state; temp_state >>= 1)
      if (temp_state & 1) sum++;
```

EUA replied:
Okay.

crankyoldgit commented on src/ir_Vestel.cpp:
```diff
@@ -0,0 +1,335 @@
+  return (( (state >> 12) & 0xFF ) == calcChecksum(state));
+}
+
+String zero_pad_str(uint8_t a){
```
This should probably be removed and use [IRWhirlpoolAc::timeToString(const uint16_t minspastmidnight)](https://github.com/markszabo/IRremoteESP8266/blob/master/src/ir_Whirlpool.cpp#L401) instead later on.
It's code re-use with code that has unit-tests and avoids us needed the `sprintf()` call.

EUA replied:
I prefer just copy whole function instead of including it...

EUA committed:
change function from snprintf to sprintf

crankyoldgit commented on src/ir_Vestel.cpp:
```diff
@@ -0,0 +1,335 @@
+#endif  // ARDUINO
+  if( remote_state.power == 0x00 ){
+        result += "Timer Command";
+        result += " - Time : " + zero_pad_str(remote_state.t_hour) + ":" + zero_pad_str(remote_state.t_minute);
```
See earlier comment at zero_pad_str():
```
result += " - Time : " + IRWhirlpoolAc::timeToString(remote_state.t_hour * 60 + remote_state.t_minute);
```

crankyoldgit commented on src/ir_Vestel.cpp:
```diff
@@ -0,0 +1,335 @@
+        result += "Timer Command";
+        result += " - Time : " + zero_pad_str(remote_state.t_hour) + ":" + zero_pad_str(remote_state.t_minute);
+	if( remote_state.t_timer_mode )
+		result += ", Timer Mode Off After " + zero_pad_str( remote_state.t_turnOffHour ) + ":" + zero_pad_str( remote_state.t_turnOffMinute*10 );
```
See earlier comment at zero_pad_str():
```
result += ", Timer Mode Off After " + IRWhirlpoolAc::timeToString(remote_state.t_turnOffHour * 60 + remote_state.t_turnOffMinute * 10);
```

Obviously, similar for others.

crankyoldgit commented on src/ir_Vestel.cpp:
```diff
@@ -0,0 +1,335 @@
+    case kVestelACFanMed:   result += " (MED)";   break;
+    case kVestelACFanHigh:  result += " (HI)";    break;
+  }
+  result += ", Sleep: "+String(getSleep()?"On":"Off");
```
Please use:
```
result += ", Sleep: ";
result += (getSleep() ? "On" : "Off");
```
as `String()` will break the unit tests as `std::string` is used in normal c++/the unit test environment.

crankyoldgit commented on src/ir_Vestel.cpp:
```diff
@@ -0,0 +1,335 @@
+
+  // Compliance
+  if (strict)
+    if( IRVestelAC::validChecksum(data_result.data) == false ) return false;
```
minor style nitpick: `if (!IRVestelAC::validChecksum(data_result.data)) return false;`

crankyoldgit commented on src/ir_Vestel.h:
```diff
@@ -0,0 +1,144 @@
+	//Remotes Timer Stack for programming AC unit for turn of after some time or self wakeup - turn off at defined time.
+	//There are no timer functions implemented.
+	//Here is the stack only for decoding and reverse enginering purposes.
+	struct{
```
Wow. This is uber neat. I didn't know about this method of packing. I'll try to use it in future!
Thumbs up!!

EUA committed:
Trying fix Travis.

crankyoldgit commented on src/ir_Vestel.cpp:
```diff
@@ -0,0 +1,335 @@
+}
+
+// Override the internal state with the new state.
+void IRVestelAC::setRaw(uint8_t* newState) {
```
You don't need to support setting it from a pointer/array. Setting it from a `uint64_t` is perfectly fine IMHO.
i.e. a straight copy/assign will do.
e.g.
```
void IRVestelAC::setRaw(uint64_t newState) {
```

EUA replied:
Well, It was uint64_t before..
Indeed I just change it to follow same pattern with other libraries while decode time.
Looks more compatible to me. Doesn't it?

```
#if DECODE_VESTEL_AC
  if (results->decode_type == VESTEL_AC) {
    IRVestelAC ac(0);
    ac.setRaw( results->state );
    description = ac.toString();
  }
#endif  // DECODE_VESTEL_AC
```

crankyoldgit replied:
See the Coolix description decode which uses `results->value`

crankyoldgit commented on src/ir_Vestel.cpp:
```diff
@@ -0,0 +1,335 @@
+
+// Return the internal state date of the remote.
+uint64_t IRVestelAC::getRaw() {
+  return remote_state.rawCode;
```
You need to do a `checksum();` here too.

EUA replied:
Checksum just fixes the CRC of the rawCode.
Since we don't send it, I don't think if need to be correct. Does it?

crankyoldgit replied:
`getRaw()` should return a fully valid code. Hence safest to force a checksum recalc before a user accesses the (new) state.

e.g.
```
IRVestelAC ac(4);
ac.setTemp(18);  // Do some arbitrary changes to a state.
sendVestelAC(ac.getRaw());  // Yes, they could use ac.send() instead, but we don't know what a user will do.
```

EUA committed:
Added std space to sprintf

crankyoldgit commented on src/ir_Vestel.cpp:
```diff
@@ -0,0 +1,335 @@
+// Return the Ion state of the A/C.
+bool IRVestelAC::getIon() { return remote_state.ion == kVestelACIon; }
+
+// Set the Wing Roaming state of the A/C.
```
Is Wing the same/similar to "Swing" mode in other A/C's? As in it causes the direction of airflow to change?
If so, could you consider using setSwing()/getSwing() etc instead to make it more consistent with the other classes?

EUA replied:
Yep. I don't looked other libs and Engish is not my native language. Can't guess such things proper :) Could rename easy.

crankyoldgit reviewed (commented):
That will do for now. I'll take a look after we get it passing Travis.

(sorry for all the nitpicks. It's really good. I'm impressed!)

crankyoldgit commented:
FYI, `clang-format -style=google -i src/ir_Vestel.cpp src/ir_Vestel.h` may help with some of the common linter errors.

crankyoldgit commented on src/ir_Vestel.h:
```diff
@@ -0,0 +1,139 @@
+const uint8_t kVestelACIon     = 4;
+const uint8_t kVestelACWing    = 0xA;
+
+union VestelACState{
```
We are getting the compiler _warnings_ when this compiles:
```
../src/ir_Vestel.h:57:9: note: offset of packed bit-field ‘VestelACState::<anonymous struct>::CRC’ has changed in GCC 4.4
   struct{
         ^
../src/ir_Vestel.h:57:9: note: offset of packed bit-field ‘VestelACState::<anonymous struct>::zero’ has changed in GCC 4.4
../src/ir_Vestel.h:76:9: note: offset of packed bit-field ‘VestelACState::<anonymous struct>::t_CRC’ has changed in GCC 4.4
   struct{
         ^
../src/ir_Vestel.h:76:9: note: offset of packed bit-field ‘VestelACState::<anonymous struct>::t_turnOffHour’ has changed in GCC 4.4
../src/ir_Vestel.h:76:9: note: offset of packed bit-field ‘VestelACState::<anonymous struct>::t_turnOnHour’ has changed in GCC 4.4
../src/ir_Vestel.h:76:9: note: offset of packed bit-field ‘VestelACState::<anonymous struct>::t_hour’ has changed in GCC 4.4
```
I **think** this may address them. Untested/not sure.

```cpp
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wpacked-bitfield-compat"
union VestelACState{
...
};
#pragma GCC diagnostic pop
```

Ref:
  https://stackoverflow.com/questions/41253759/ignoring-note-offset-of-packed-bit-field-without-using-wno-packed-bitfield-co
  https://www.gnu.org/software/gcc/gcc-4.4/changes.html

EUA replied:
Well, not warnings but _notes_ this time. :)

I add those pragma messages but doesn't kill the notes at Arduino GCC4.4

EUA committed:
Added Vestel A/C decode support to IRRecvDumpV2 Example.
Fixed some errors.

EUA committed:
Removed sprintf and replaced with timeToString fx.

EUA committed:
Added CRC fix before GetRaw();

EUA commented:
9 am here. Will continue next day. :+1:

crankyoldgit committed:
Cleanup cpplint issues.
i.e. Ran `clang-gormat -style=google` over code.

crankyoldgit committed:
Update example code to handle Vestel AC.
- Add a setRaw() that accepts a uint64_t.
- Should correct travis build issues now.

crankyoldgit committed:
Remove duplicate code (timeToString())
Use IRHaierAC::timeToString() instead.

crankyoldgit committed:
Add basic unit tests for Vestel A/C.
- Add some extra fan speed modes.

crankyoldgit commented:
@EUA I've done some cleanup, fixed a few things, added support for the code examples, and added basic unit tests for it.

Doing some reading on the structure bit packing, it's "implementation dependent" which means we really shouldn't use it. e.g. Different compilers and architectures could produce problems.

Also, when writing the unit tests, I noticed there are no class methods to set any of the clock/timer options. They are optional, but the should be added at some time.

Can you supply some other values for valid codes (and what they mean) so I can add some more tests?
Also, a rawData entry would also be good for a unit test for the decoder too.

EUA commented:
About packing, yes this is old and not a meaningful thing to talk in 2019. Sad to hear it again. I want to talk about it (as a msc computer engineer.)
Some people just love to talk about bitfield ops dangers but none of them explain about why they want to use such an "half feature set" compiler. If we don't use such a features why they are exists? I even hear some people against C++ inheritances due similar concerns (compatibility) many years ago... So why many people trying to add/implement new feature sets under C++1x since many of compilers don't gonna support them?
Also many many program/library uses inline assembly for x86 or ARM without feeling such a concerns. Since this library is already depend to Arduino and ESP and you have to use xtensa compiler due no alternative of it, I don't think if we need to worry about any compatibility issues at this point, right? If we are working on a project like "bash", yes it's better to put compatibility to the first place..
Well, GCC & LLVM supports BitField ops... Also remaining compilers probably will work ... if they are implemented proper. BitField ops are **easy** to implemented on a compiler. And every CPU architecture that has a "shift register" in it will support BitFields. Do you know any CPU without it?
BitField Ops is not an evil that we need to fear. For me it's old and good _C art._
And the best part is _**you can visualize the protocol stack structure in your head at the first look.**_ Which is priceless in coding. If there are any bugs appears later in x compiler at y architecture, I believe it's easy to fix since coder see the whole operation. It's easy to patch if you understand the structure.

Sorry. If you don't like or can't use it at this point, you can simply reject the code. I don't gonna use some other "compatible" method for implement it. I just picked the best. 

For timer features, I actually don't see any possible benefit of such an implementation. 
Because it's easier and better to program MCU instead of "programming MCU to program A/C unit".

Well, Vestel AC has off-timer (30 min, 1hr,2hr,3hr 5hr). Also has option to program turn on and turn off selected time but in 10 minutes interval. If you really want I could add it. 

I can send IRecvDumpV2 outputs later. Can I paste here or???

EUA commented:
Now realized that... Code flipped some how at IRrecvDumpV2...
I return with git to older state and... It still same.
I mean:

```
Encoding  : VESTEL_AC
Code      : F441E001FEE201 (56 bits)
```

becomes
```
Encoding  : VESTEL_AC
Code      : 01E2FE01E041F4 (56 bits)
```
Everything works proper, nothing changed at all but I swear yesterday (and my whole development) it's F441E001FEE201...
I check reverse engineering codes and. It was in form of: F441E001FEE201...
What happening here?? I don't have any idea about whats going on here.

EUA commented:
I am really scared... My perception of real is altered for a while. Just like traveling parallel universes...

But I approximate and find the problem.
I developed using library v2.5.3.
But for pushing, I switched v2.5.4 from git. And it become flipped. Not realized due reverse engineering competed.

Than, this thing became a bug.

```
IRrecvDumpV2 is now running and waiting for IR input on Pin 14
Timestamp : 000014.488
Encoding  : VESTEL_AC
Code      : 0112FF010041F4 (56 bits)
Mesg Desc.: Power: On, Mode: 4 (HEAT), Temp: 16C, Fan: 1 (AUTO), Sleep: Off, Turbo: Off, Ion: On, Swing: Off
Library   : v2.5.4

Raw Timing[115]:
   +  3098, -  9080,    +   548, -  1538,    +   526, -   492,    +   526, -   468, 
   +   524, -   468,    +   526, -   468,    +   550, -   466,    +   526, -   466, 
   +   526, -   504,    +   540, -   466,    +   526, -  1538,    +   526, -   466, 
   +   526, -   466,    +   552, -  1540,    +   522, -   466,    +   526, -   492, 
   +   526, -   544,    +   526, -  1536,    +   526, -  1536,    +   552, -  1536, 
   +   526, -  1536,    +   552, -  1536,    +   552, -  1536,    +   526, -  1536, 
   +   526, -  1574,    +   542, -  1536,    +   526, -   492,    +   526, -   466, 
   +   526, -   494,    +   524, -   468,    +   524, -   468,    +   526, -   492, 
   +   526, -   502,    +   540, -   468,    +   524, -   494,    +   524, -   468, 
   +   526, -   468,    +   524, -   468,    +   526, -   492,    +   526, -   468, 
   +   524, -   520,    +   524, -  1538,    +   524, -   468,    +   524, -   468, 
   +   524, -   468,    +   524, -   468,    +   524, -   468,    +   524, -  1538, 
   +   524, -   506,    +   538, -   468,    +   524, -   468,    +   524, -  1538, 
   +   524, -   468,    +   550, -  1538,    +   550, -  1538,    +   524, -  1538, 
   +   534, -  1528,    +   544

uint16_t rawData[115] = {3098, 9080,  548, 1538,  526, 492,  526, 468,  524, 468,  526, 468,  550, 466,  526, 466,  526, 504,  540, 466,  526, 1538,  526, 466,  526, 466,  552, 1540,  522, 466,  526, 492,  526, 544,  526, 1536,  526, 1536,  552, 1536,  526, 1536,  552, 1536,  552, 1536,  526, 1536,  526, 1574,  542, 1536,  526, 492,  526, 466,  526, 494,  524, 468,  524, 468,  526, 492,  526, 502,  540, 468,  524, 494,  524, 468,  526, 468,  524, 468,  526, 492,  526, 468,  524, 520,  524, 1538,  524, 468,  524, 468,  524, 468,  524, 468,  524, 468,  524, 1538,  524, 506,  538, 468,  524, 468,  524, 1538,  524, 468,  550, 1538,  550, 1538,  524, 1538,  534, 1528,  544};  // VESTEL_AC
uint8_t state[7] = {0x01, 0x12, 0xFF, 0x01, 0x00, 0x41, 0xF4};
```
auto_analyse_raw_data.py  output still show the old notation:
```
Found 115 timing entries.
Potential Mark Candidates:
[3098, 552]
Potential Space Candidates:
[9080, 1574, 544]

Guessing encoding type:
Looks like it uses space encoding. Yay!

Guessing key value:
kHdrMark   = 3098
kHdrSpace  = 9080
kBitMark   = 530
kOneSpace  = 1538
kZeroSpace = 478

Decoding protocol based on analysis so far:

kHdrMark+kHdrSpace+10000000010010001111111110000000000000001000001000101111
  Bits: 56
  Hex:  0x8048FF8000822F (MSB first)
        0xF4410001FF1201 (LSB first)
  Dec:  36109059220341295 (MSB first)
        68751362606699009 (LSB first)
  Bin:  0b10000000010010001111111110000000000000001000001000101111 (MSB first)
        0b11110100010000010000000000000001111111110001001000000001 (LSB first)

Total Nr. of suspected bits: 56
```

EUA committed:
Reverted last changes about AutoFan mode which doesn't exists.

EUA commented:
Auto modes exist but Fan mode is different than that. Doesn't make heat or cold. 
I find that not implemented an usable function for use auto modes.
I will add a function for it asap.

about 
`  remote_state.temp = new_temp - kVestelACMinTempH;`
Well AC protocol does not take kVestelACMinTempH as a reference because there are 2 min temp available at unit. One of them 18 (for cooling), other is 16 (for heating).
I think it's better to leave it as 16. Because using MinTemp instead of 16 will work on code but logically it's a mistake since there are no a real correlation between them actually. Do they have correlation for you? If you think like that. It's OK also.

crankyoldgit commented:
> About packing, yes this is old and not a meaningful thing to talk in 2019. Sad to hear it again. I want to talk about it (as a msc computer engineer.)
> Some people just love to talk about bitfield ops dangers but none of them explain about why they want to use such an "half feature set" compiler. If we don't use such a features why they are exists? I even hear some people against C++ inheritances due similar concerns (compatibility) many years ago... So why many people trying to add/implement new feature sets under C++1x since many of compilers don't gonna support them?
> Also many many program/library uses inline assembly for x86 or ARM without feeling such a concerns. Since this library is already depend to Arduino and ESP and you have to use xtensa compiler due no alternative of it, I don't think if we need to worry about any compatibility issues at this point, right? If we are working on a project like "bash", yes it's better to put compatibility to the first place..
> Well, GCC & LLVM supports BitField ops... Also remaining compilers probably will work ... if they are implemented proper. BitField ops are **easy** to implemented on a compiler. And every CPU architecture that has a "shift register" in it will support BitFields. Do you know any CPU without it?
> BitField Ops is not an evil that we need to fear. For me it's old and good _C art._
> And the best part is _**you can visualize the protocol stack structure in your head at the first look.**_ Which is priceless in coding. If there are any bugs appears later in x compiler at y architecture, I believe it's easy to fix since coder see the whole operation. It's easy to patch if you understand the structure.
> 
> Sorry. If you don't like or can't use it at this point, you can simply reject the code. I don't gonna use some other "compatible" method for implement it. I just picked the best.

I read up on a number of issues about it last night. The BitField operations are endian-dependant. i.e. Different architectures can cause it to "flip" etc. The AVR/ESP/x86 all share the same endian-ness so we are fine there. However, I did read the one of Microsoft's compiler/IDE environments does treat bitfield packing differently to GCC on the same arch. i.e. produces different bit packing. I do know some people/project that do use this library with that compiler env. so ultimately we should probably change it (not for this PR). Also, we have no control of which architecture the CI/Travis uses. As the unit tests compile outside of the AVR cross-compiler, we do need to cater for that. Currently, they are using x86 it seems so all is fine for now.

I'll probably re-code it to the correct bit-shifting/masking in a later PR.

crankyoldgit commented on src/ir_Vestel.h:
```diff
@@ -0,0 +1,146 @@
+    uint8_t turbo_sleep_normal : 4;  //  normal 0x1, sleep 0x3, turbo 0x7
+    uint8_t zero : 8;                // 0x00
+    uint8_t temp : 4;                // temp-16 degree Celcius
+    uint8_t fan : 4;  // auto 0x1, low 0x5, mid 0x9, high 0xB, 0xD auto hot, 0xC
```
This is where I got the values for AutoHot and AutoCool from. AutoHot is also the value that was in your _default_ state in the `stateReset()` function. Is this comment wrong?

crankyoldgit replied:
See https://travis-ci.org/markszabo/IRremoteESP8266/jobs/485705512#L7993 & https://travis-ci.org/markszabo/IRremoteESP8266/jobs/485705512#L8004

It now results in "Fan: 13 (UNKNOWN)", we should really name that value.

EUA replied:
Yeah. I just think there are one Fan modes exist on auto due code reversal for a while.

EUA committed:
Revert last revert. AutoFanModes are exitst.
Added AutoMode function for easy switch.

crankyoldgit commented:
> Everything works proper, nothing changed at all but I swear yesterday (and my whole development) it's F441E001FEE201...
> I check reverse engineering codes and. It was in form of: F441E001FEE201...
> What happening here?? I don't have any idea about whats going on here.

My guess is you used the wrong bitordering in the `MSBfirst` arguement of `matchData()` in the decode routine. or you need to reverse the bit ordering before assigning it to `result.value`.
Note: There is a difference when assembly an array of `uint_8`s vs. an entire `uint64_t` if you have a different bit ordering for each point.

FYI, nothing changed from 2.5.3 to 2.5.4 that would have affected this. My guess is you were looking at it character by character, than as a whole uint64_t.

This is why unit tests are good. ;-)

crankyoldgit commented:
> `remote_state.temp = new_temp - kVestelACMinTempH;`
> Well AC protocol does not take kVestelACMinTempH as a reference because there are 2 min temp available at unit. One of them 18 (for cooling), other is 16 (for heating).
> I think it's better to leave it as 16. Because using MinTemp instead of 16 will work on code but logically it's a mistake since there are no a real correlation between them actually. Do they have correlation for you? If you think like that. It's OK also.

I typically prefer to use a named constant than rather plain number, especially if it used in more than one place. 16 is the value which is the temp offset. kVestelACMinTempH is in effect that value. i.e. it is the lowest temp that can be represented (i.e. 0b0000 in the message).
It's not a hard & fast rule, but in this particular case, it is relatively the same in all the other A/C encoder/decoders.

TL;DR: I don't really care. :)

EUA committed:
Added Timer and WakeupTime & TurnOffTime functions.

crankyoldgit commented on src/ir_Vestel.cpp:
```diff
@@ -175,6 +175,43 @@ void IRVestelAC::setAuto(const int8_t autoLevel) {
+    }
+  else{
+    remote_state.t_turnOffHour=minutes/60;
+    remote_state.t_turnOffMinute=minutes/10;
```
I think you mean:
`remote_state.t_turnOffMinute = (minutes % 60) / 10;`

EUA committed:
Fixed timer.

crankyoldgit commented on src/ir_Vestel.cpp:
```diff
@@ -175,6 +175,43 @@ void IRVestelAC::setAuto(const int8_t autoLevel) {
+// Set AC's wake up time. Takes time in minute.
+void IRVestelAC::setWakeupTime(const uint16_t minutes){
+  remote_state.t_turnOnHour=minutes/60;
+  remote_state.t_turnOnMinute=((minutes/60)*minutes-minutes)/10; // Register is 3 bit and can only hold minute/10
```
remote_state.t_turnOnMinute= (minutes % 60) / 10;  // Register is 3 bit and can only hold minute/10

EUA replied:
Yes, we have a "%" right...
In this state of mind. I think I really need to sleep. 7:00am here

crankyoldgit replied:
Get some sleep. This can wait. (and I can make commits without having to merge ;-)

crankyoldgit commented on src/ir_Vestel.cpp:
```diff
@@ -175,6 +175,43 @@ void IRVestelAC::setAuto(const int8_t autoLevel) {
+// Set AC's turn off time. Takes time in minute.
+void IRVestelAC::setTurnOffTime(const uint16_t minutes){
+  remote_state.t_turnOffHour=minutes/60;
+  remote_state.t_turnOffMinute=((minutes/60)*minutes-minutes)/10; // Register is 3 bit and can only hold minute/10
```
ditto.

crankyoldgit committed:
VESTEL_AC does not use State.
Fix incorrect determination of if VESTEL uses uint8[] for state.
It doesn't. It uses uint64_t instead.
Should fix incorrect reporting in IRrecvDumpV2.

Add unit tests to ensure it is correct.

* Fix a metric tonne of linter issues.

EUA committed:
Fixed % at timers.

EUA committed:
Typo

crankyoldgit commented on src/ir_Vestel.cpp:
```diff
@@ -0,0 +1,436 @@
+std::string IRVestelAC::toString() {
+  std::string result = "";
+#endif  // ARDUINO
+  if (remote_state.power == 0x00) {
```
Is this correct? I don't see anything that ever sets this to `0`. only `0xF` or `0xC`

EUA replied:
Yes. This is main difference to identify normal command stack or time stack.

If remote_state.power byte is not set, that means we are decoding time command.

I don't add any reset for that bits since not planning to implement timing functions.
I reset it using t_not_used.
But using power=0x00 might be better idea.

crankyoldgit replied:
My vote is to use `power` rather than `not_used` as clearly it is used.

Can you give me a code for a timer message so I can build some unit tests around it and fix toString() to reflect that.

EUA replied:
I just add new remote state for time.. So switching is not needed anymore.

Here is the some codes. "Mesg Desc" shows the time command function :)

```
Timestamp : 000008.851
Encoding  : VESTEL_AC
Code      : 2D6570B8EE201 (56 bits)
Mesg Desc.: Time Command - Time : 5:45, Turn On: 14:00, Turn Off: 23:00
Library   : v2.5.4

Raw Timing[115]:
   +  3022, -  9080,    +   546, -  1536,    +   526, -   466,    +   526, -   492, 
   +   526, -   468,    +   526, -   492,    +   524, -   468,    +   524, -   494, 
   +   524, -   504,    +   540, -   492,    +   524, -  1538,    +   526, -   468, 
   +   524, -   492,    +   526, -   466,    +   552, -  1536,    +   526, -  1536, 
   +   526, -  1570,    +   542, -   492,    +   524, -  1538,    +   550, -  1538, 
   +   524, -  1536,    +   526, -   494,    +   524, -   466,    +   526, -   468, 
   +   524, -  1574,    +   540, -  1536,    +   550, -  1536,    +   526, -   468, 
   +   550, -  1536,    +   526, -   492,    +   526, -   468,    +   524, -   492, 
   +   526, -   518,    +   526, -  1536,    +   552, -  1536,    +   550, -  1536, 
   +   526, -   494,    +   550, -  1538,    +   526, -   492,    +   524, -  1538, 
   +   526, -   504,    +   540, -   466,    +   526, -  1536,    +   526, -  1536, 
   +   526, -   468,    +   550, -  1538,    +   524, -   468,    +   524, -  1538, 
   +   550, -  1574,    +   540, -   468,    +   550, -  1538,    +   526, -   492, 
   +   524, -   468,    +   526, -   466,    +   526, -   468,    +   524, -   494, 
   +   524, -   468,    +   546

uint16_t rawData[115] = {3022, 9080,  546, 1536,  526, 466,  526, 492,  526, 468,  526, 492,  524, 468,  524, 494,  524, 504,  540, 492,  524, 1538,  526, 468,  524, 492,  526, 466,  552, 1536,  526, 1536,  526, 1570,  542, 492,  524, 1538,  550, 1538,  524, 1536,  526, 494,  524, 466,  526, 468,  524, 1574,  540, 1536,  550, 1536,  526, 468,  550, 1536,  526, 492,  526, 468,  524, 492,  526, 518,  526, 1536,  552, 1536,  550, 1536,  526, 494,  550, 1538,  526, 492,  524, 1538,  526, 504,  540, 466,  526, 1536,  526, 1536,  526, 468,  550, 1538,  524, 468,  524, 1538,  550, 1574,  540, 468,  550, 1538,  526, 492,  524, 468,  526, 466,  526, 468,  524, 494,  524, 468,  546};  // VESTEL_AC 2D6570B8EE201
uint64_t data = 0x2D6570B8EE201;
```

```
Timestamp : 000053.163
Encoding  : VESTEL_AC
Code      : 2EA50018F3201 (56 bits)
Mesg Desc.: Time Command - Time : 5:46, Timer Mode Off After 03:00
Library   : v2.5.4

Raw Timing[115]:
   +  3100, -  9076,    +   550, -  1534,    +   528, -   490,    +   528, -   464, 
   +   528, -   490,    +   528, -   464,    +   528, -   490,    +   528, -   466, 
   +   526, -   502,    +   542, -   464,    +   554, -  1534,    +   528, -   490, 
   +   528, -   464,    +   552, -  1534,    +   528, -  1534,    +   528, -   464, 
   +   528, -   526,    +   542, -  1534,    +   552, -  1534,    +   528, -  1536, 
   +   526, -  1534,    +   528, -   466,    +   528, -   464,    +   528, -   492, 
   +   526, -  1572,    +   542, -  1534,    +   528, -   490,    +   528, -   490, 
   +   526, -   466,    +   528, -   490,    +   528, -   464,    +   528, -   466, 
   +   526, -   502,    +   542, -   464,    +   528, -   490,    +   528, -   490, 
   +   528, -   466,    +   552, -  1534,    +   528, -   490,    +   528, -  1534, 
   +   528, -   526,    +   542, -   490,    +   554, -  1534,    +   528, -   466, 
   +   552, -  1534,    +   528, -   464,    +   552, -  1536,    +   552, -  1534, 
   +   528, -  1570,    +   544, -   490,    +   528, -  1534,    +   528, -   466, 
   +   526, -   490,    +   528, -   466,    +   528, -   490,    +   528, -   490, 
   +   526, -   466,    +   550

uint16_t rawData[115] = {3100, 9076,  550, 1534,  528, 490,  528, 464,  528, 490,  528, 464,  528, 490,  528, 466,  526, 502,  542, 464,  554, 1534,  528, 490,  528, 464,  552, 1534,  528, 1534,  528, 464,  528, 526,  542, 1534,  552, 1534,  528, 1536,  526, 1534,  528, 466,  528, 464,  528, 492,  526, 1572,  542, 1534,  528, 490,  528, 490,  526, 466,  528, 490,  528, 464,  528, 466,  526, 502,  542, 464,  528, 490,  528, 490,  528, 466,  552, 1534,  528, 490,  528, 1534,  528, 526,  542, 490,  554, 1534,  528, 466,  552, 1534,  528, 464,  552, 1536,  552, 1534,  528, 1570,  544, 490,  528, 1534,  528, 466,  526, 490,  528, 466,  528, 490,  528, 490,  526, 466,  550};  // VESTEL_AC 2EA50018F3201
uint64_t data = 0x2EA50018F3201;
```

EUA committed:
Fixed internal state change for time mode.

EUA commented:
Yes verified that code reversal is fixed :) Happy now.

I implemented date timer setting but do not verify if they are work.
Since decoder stack is working and detect real remote commands, I believe you can check it.

Only current pitfall is using same VestelACState remote_state for both type of commands.
Those are completely different.
For use time stack, first we need to reset remote_state using **setTimer(0)**.
Otherwise it corrupts the remote_state

Or we can use another remote_time_state for timing operations.
I don't think on wide because not planning to implement time stack functions...

crankyoldgit commented:
Ah. So there is an entirely different 56bit message layout if it's timer related vs. a change mode/temp/etc message?  If so, Yuck.

EUA commented:
They call it as **hug** in Linux kernel anymore... :)

Yes. Only CRC bits and t_footer are in same place :)

Might be it's better to use different internal state named like **remote_time_state** for timing operations. Will add/change it tomorrow. Than it's done...

EUA committed:
Added remote_time_state for timing ops.

EUA committed:
Remote timer state set to zero on constractor.

EUA committed:
Fix for travis.

crankyoldgit committed:
Add better support for time(r) related Vestel commands.
- Make sure IRVestelAC::send() knows which state to send.
  * Set the type of command to send to be the type of the last method type used.
    e.g. If a time(r) function was used last, send a Time message, otherwise
         a normal mode message is sent.
- Clean up/standardise the timer human readable text.
- Add missing timer get() functions. Rename set() functions to common names.
- Add unit tests to cover pretty much all the missing and new class methods.
- Add real capture data for unit tests.

crankyoldgit commented:
This has been included in the new v2.5.5 release of the library.
